WordPress/src/test/java/org/wordpress/android/ui/blaze/blazecampaigns/CampaignDetailViewModelTest.kt

```diff
@@ -0,0 +1,152 @@
+        whenever(accountStore.account.userName).thenReturn(null)
+
+        val uiModels = mutableListOf<CampaignDetailUIModel>()
+        val actionEvents = mutableListOf<BlazeActionEvent>()
+        testWithData(actionEvents, uiModels) {
+            viewModel.start(1, CampaignDetailPageSource.DASHBOARD_CARD)
+
+            assertThat(actionEvents.last()).isInstanceOf(BlazeActionEvent.FinishActivityWithMessage::class.java)
+        }
+    }
+
+    @Test
+    fun `given invalid accessToken, when start is called, then FinishActivityWithMessage is posted`() {
+        whenever(accountStore.accessToken).thenReturn(null)
+
+        val uiModels = mutableListOf<CampaignDetailUIModel>()
+        val actionEvents = mutableListOf<BlazeActionEvent>()
+        testWithData(actionEvents, uiModels) {
+            viewModel.start(1, CampaignDetailPageSource.DASHBOARD_CARD)
+
+            assertThat(actionEvents.last()).isInstanceOf(BlazeActionEvent.FinishActivityWithMessage::class.java)
+        }
+    }
+    @Test
+    fun `given valid account info, when start is called, then url is built`() {
+        val domain = "test.wordpress.com"
+        val postData = "postdata"
+        whenever(blazeFeatureUtils.getAuthenticationPostData(any(), any(), any(), any(), any())).thenReturn(postData)
+
+        val uiModels = mutableListOf<CampaignDetailUIModel>()
+        val actionEvents = mutableListOf<BlazeActionEvent>()
+        testWithData(actionEvents, uiModels) {
+            viewModel.start(1, CampaignDetailPageSource.DASHBOARD_CARD)
+
+            val uiModel = uiModels.last()
+            assertThat(uiModel.url).isNotEmpty
+            assertThat(uiModel.url).contains(domain)
+            assertThat(uiModel.addressToLoad).isNotEmpty
+            assertThat(uiModel.addressToLoad).contains(postData)
+        }
+    }
+    @Test
+    fun `when onRedirectToExternal browser is requested, then launch external browser action event is posted`() {
+        val uiModels = mutableListOf<CampaignDetailUIModel>()
+        val actionEvents = mutableListOf<BlazeActionEvent>()
+        val url = "https://external.url.com"
+
+        testWithData(actionEvents, uiModels) {
+            viewModel.onRedirectToExternalBrowser(url)
+
+            val actionEvent = actionEvents.last()
+            assertThat(actionEvent).isInstanceOf(BlazeActionEvent.LaunchExternalBrowser::class.java)
+            assertThat((actionEvent as BlazeActionEvent.LaunchExternalBrowser).url).isEqualTo(url)
+        }
+    }
+    @Test
+    fun `when on web view error handled, then finish activity with message event is posted`() {
+        val uiModels = mutableListOf<CampaignDetailUIModel>()
+        val actionEvents = mutableListOf<BlazeActionEvent>()
+        val errorMessageResId = R.string.blaze_campaign_detail_error
+
+        testWithData(actionEvents, uiModels) {
+            viewModel.onWebViewError()
```

Comment:
Nice. 👏🏼 . Adding the case for webview error as well

Reply:
ikr ... also added this to promote in the next PR. :)